File: algorithms/java/109_convert_sorted_list_to_binary_search_tree/Convert.java
public class Convert {
    
    // 2. recursion
    public TreeNode sortedListToBST(ListNode head) {
        return construct(head);
    }
    private TreeNode construct(ListNode head) {
        if (head == null) return null;
        if (head.next == null) return new TreeNode(head.val);
        ListNode mid = findMid(head);
        TreeNode root = new TreeNode(mid.val);
        root.left = construct(head);
        root.right = construct(mid.next);
        return root;
    }
    private ListNode findMid(ListNode head) {
        ListNode slow = head;
        ListNode prev = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
		prev->next = null;
        return slow;
    }

    // 3. use a global pointer to inorder traverse linkedlist pointing to the root
    public TreeNode sortedListToBST1(ListNode head) {
        if (head == null) return null;
        int len = getLength(head);
        ListNode[] cur = new ListNode[]{head};
        return construct(0, len - 1, cur);
    }
    private TreeNode construct(int start, int end, ListNode[] cur) {
        if (start > end) return null;
        int mid = start + (end - start) / 2;
        TreeNode left = construct(start, mid - 1, cur);
        TreeNode root = new TreeNode (cur[0].val);
        cur[0] = cur[0].next;
        root.right = construct(mid + 1, end, cur);
        root.left = left;
        return root;
    }
    private int getLength(ListNode head) {
        int len = 0;
        while (head != null) {
            head = head.next;
            len++;
        }
        return len;
    }
}

File: algorithms/cpp/109_convert_list_bst/Convert.cpp
class Convert {
public:
	// O(nlogn) recursion
    TreeNode* sortedListToBST(ListNode* head) {
        return construct(head);
    }
    TreeNode* construct(ListNode* head) {
        if (head == NULL)
            return NULL;
        if (head->next == NULL)
            return new TreeNode(head->val);
        ListNode* mid = find_mid(head);
        TreeNode* root = new TreeNode(mid->val);
        root->left = construct(head);
        root->right = construct(mid->next);
        return root;
    }
    ListNode* find_mid(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* prev = head;
        while (fast && fast->next) {
            prev = slow;
            slow = slow->next;
            fast = fast->next->next;
        }
        prev->next = NULL;
        return slow;
    }

	// O(n) recursion
	TreeNode* sortedListToBST1(ListNode* head) {
        if (head == NULL)
            return NULL;
        int len = get_len(head);
        ListNode* cur = head;
        return construct1(0, len - 1, cur);
    }
    TreeNode* construct1(int start, int end, ListNode*& cur) {
        if (start > end)
            return NULL;
        int mid = start + (end - start) / 2;
        TreeNode* left = construct1(start, mid - 1, cur);
        TreeNode* root = new TreeNode(cur->val);
        cur = cur->next;
        root->right = construct1(mid + 1, end, cur);
        root->left = left;
        return root;
    }
    int get_len(ListNode* head) {
        int len = 0;
        while (head) {
            head = head->next;
            ++len;
        }
        return len;
    }
};
